File: 2017_July/Day_Two/Nuts_and_Bolts/session1/exercises/mpi_collatz.cpp
///////////////////////////////////////////////////////////////////
//
//          MPI Exercise:  Collatz Sequence
//
//          Modify the program below so that it reports the number
//          with the largest collatz sequence length, and its length,
//          occuring in the interval [1, N Million] where N is the 
//          number of MPI ranks.
#include <stdio.h>
#include <mpi.h>
int collatz_length(int n) {
    int i;
    int length;
    length=1;
    i = n;
    while (i > 1) {
        length +=1;
        if ( ( i%2 ) ==0 ){
            i = i/2;
        } else {
            i = 3*i+1;
        }
    }
    return length;
}

int main(int argc, char** argv) {
    int ntests;
    int jstart;
    int jend;
    int jlength;
    int jmax;              // The maximum number whose sequence-length is calculated by any processor
    int njlocal;           // How many numbers each process checks the sequence length for.
    int max_length_local;  // The maximum sequence length calculated locally
    int max_length_global; // The maximum sequence length calculated globally
    int max_num_local;     // The number with the longest sequence-length found locally
    int max_num_global;    // The number possessing the global max sequence length
    int my_rank;
    int num_proc;

    MPI_Init(NULL, NULL);
    MPI_Comm_size(MPI_COMM_WORLD, &num_proc);
    MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

    njlocal = 1000000;
    jmax = njlocal*num_proc;

    if (my_rank == 0)
    {
        printf("  Calculating Collatz-sequence lengths over the interval [1,%d] using %d processes.\n",jmax, num_proc);
    }

    // (1) Change jstart and jend so that all rank compute a sequence lengths for numbers
    //     in a unique portion of the interval [1,jmax]
    jstart = 1;
    jend = jmax;

    max_length_local=1;
    max_num_local=1;
    for (int j=jstart; j<jend; ++j) {
        jlength = collatz_length(j);
        if (jlength > max_length_local){
            max_num_local=j;
            max_length_local=jlength;
        }
    }
    max_length_global=0;
    max_num_global=0;
    // (2) Use an Allreduce to get the global maximum of the sequence lengths
    printf(" The longest Collatz-sequence in the interval [ 1, %d ] has length %d.\n", jmax,max_length_global);

    // (3;  A little bit trickier...) Use an Allreduce to assign the number associated with that sequence length to max_num_global
    printf(" This sequence-length occurs for the number %d.\n", max_num_global);
    ierr = MPI_Finalize();
}

File: 2017_July/Day_Two/Nuts_and_Bolts/session1/exercises/mpi_collatz.f90
!/////////////////////////////////////////////////////////////////
!
!          MPI Exercise:  Collatz Sequence
!
!          Modify the program below so that it reports the number
!          with the largest collatz sequence length, and its length,
!          occuring in the interval [1, N Million] where N is the 
!          number of MPI ranks.
PROGRAM MAIN 
    IMPLICIT NONE
    include "mpif.h"

    Integer :: j
    Integer :: jstart
    Integer :: jend
    Integer :: jlength
    Integer :: jmax              ! The maximum number whose sequence-length is calculated by any processor
    Integer :: njlocal           ! How many numbers each process checks the sequence length for.
    Integer :: max_length_local  ! The maximum sequence length calculated locally
    Integer :: max_length_global ! The maximum sequence length calculated globally
    Integer :: max_num_local     ! The number with the longest sequence-length found locally
    Integer :: max_num_global    ! The number possessing the global max sequence length
    Integer :: my_rank
    Integer :: num_proc
    Integer :: ierr



    Call MPI_INIT( ierr )
    Call MPI_Comm_size(MPI_COMM_WORLD, num_proc,ierr)
    Call MPI_Comm_rank(MPI_COMM_WORLD, my_rank,ierr)
    

    njlocal = 1000000
    jmax = njlocal*num_proc

    
    If (my_rank .eq. 0) Then
       Write(6,'(a,i0,a,i0,a)')"  Calculating Collatz-sequence lengths over the interval [1,", jmax, &
                "] using ", num_proc, " processes."
    Endif
    ! (1) Change jstart and jend so that all rank compute a sequence lengths for numbers
    !     in a unique portion of the interval [1,jmax]
    jstart = 1
    jend = jmax

    max_length_local=1
    max_num_local=1
    Do j = jstart, jend
        jlength = collatz_length(j)
        If (jlength .gt. max_length_local) Then 
            max_num_local=j
            max_length_local=jlength
        Endif

    Enddo
    max_length_global=0
    max_num_global=0





    ! (2) Use an Allreduce to get the global maximum of the sequence lengths
    Write(6,'(a,i0,a,i0,a)') "  The longest Collatz-sequence in the interval [ 1,", jmax, &
            "] has length ",max_length_global, "."
    ! (3;  A little bit trickier...) Use an Allreduce to assign the number associated with that sequence length to max_num_global
    Write(6,'(a,i0,a)')"  This sequence-length occurs for the number ",max_num_global, "."


    Call MPI_Finalize(ierr)
Contains

    Function collatz_length(n) result(length)
        Integer, Intent(In) :: n
        Integer :: i, length
        length = 1
        i = n
        Do While (i .gt. 1)
            length = length+1
            If ( MOD(i,2) .eq. 0) THEN
                i = i/2
            Else
                i = 3*i+1
            Endif
        Enddo
        
    End Function collatz_length



END PROGRAM MAIN

